File: Demos/CPP/VCL/QPlugins/FormService/hostmain.cpp
// ---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "hostmain.h"
#include "qplugins_loader_lib.hpp"
// ---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TForm1 *Form1;

// ---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner) : TForm(Owner)
{
	String APath = ExtractFilePath(Application->ExeName);
	// 缓存下PluginsManager，减少重复调用，毕竟是个函数:)
	_di_IQPluginsManager AMgr = PluginsManager();
	AMgr->Loaders->Add(ToInterface<IQLoader>(new TQBPLLoader(APath, L".bpl")));
	AMgr->Loaders->Add(ToInterface<IQLoader>(new TQDLLLoader(APath, L".dll")));
	AMgr->Start();
}

// ---------------------------------------------------------------------------
void __fastcall TForm1::Button1Click(TObject *Sender)
{
	wchar_t *DockRoot = L"/Services/Docks/Forms";
	_di_IQServices AServices;
	if (HasInterface(PluginsManager()->ByPath(DockRoot), AServices))
	{
		_di_IQFormService AFormService;
		for (int i = 0; i < AServices->Count; i++)
		{
			if (HasInterface(AServices->Items[i], AFormService))
			{
				if (!AFormService->IsMultiInstance())
				{
					TTabSheet *APage = new TTabSheet(PageControl1);
					APage->PageControl = PageControl1;
					APage->Caption =ToInterface<IQService>(AFormService)->Name;
					APage->Tag = (IntPtr)AFormService.operator->();
					AFormService->DockTo((UIntPtr)APage->Handle,
						APage->ClientRect);
					TQFormEvents AEvents =
					{NULL};
					AEvents.OnFree = DoDockChildFree;
					AFormService->HookEvents(AEvents);
				}
			}
		}
	}
}

// ---------------------------------------------------------------------------
void __fastcall TForm1::DoDockChildFree(_di_IQFormService AService)
{
	IQFormService *AFormService = AService;
	for (int i = 0; i < PageControl1->PageCount; i++)
	{
		if (((void *)PageControl1->Pages[i]->Tag) == AFormService)
		{
			FreeObject(PageControl1->Pages[i]);
			break;
		}
	}
}

void __fastcall TForm1::FormDestroy(TObject * Sender)
{
	for (int i = 0; i < PageControl1->PageCount; i++)
	{
		((IQFormService*)PageControl1->Pages[i]->Tag)->UnhookEvents();
	}
}
// ---------------------------------------------------------------------------

File: Demos/Delphi/VCL/QPlugins/Thirdpart/hostmain.pas
unit hostmain;

interface

uses
  Winapi.Windows, Winapi.Messages, System.SysUtils, System.Variants,
  System.Classes, Vcl.Graphics,
  Vcl.Controls, Vcl.Forms, Vcl.Dialogs, qplugins,qplugins_base, qplugins_formsvc,
  qplugins_vcl_formsvc, qplugins_loader_lib, Vcl.StdCtrls,
  Vcl.ExtCtrls;

type
  TForm1 = class(TForm)
    Panel1: TPanel;
    Button1: TButton;
    procedure Button1Click(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure FormClose(Sender: TObject; var Action: TCloseAction);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  Form1: TForm1;

implementation

{$R *.dfm}

procedure TForm1.Button1Click(Sender: TObject);
var
  AService: IQFormService;
begin
  if Supports(PluginsManager.ByPath('Services/Docks/Forms/VSTForm'),
    IQFormService, AService) then
  begin
    AService.DockTo(Panel1.Handle, Panel1.ClientRect);
  end;
end;

procedure TForm1.FormClose(Sender: TObject; var Action: TCloseAction);
begin
// 注意：由于 DLL 中的服务使用的 VirtualStringTree 创建了后台线程
//如果不先注销后台服务的话，应用程序退出时，会由于 Delphi 自身实现的问题
//造成程序无法退出。在这里先移除注册，然后再正常关闭就好了
  UnregisterServices('Services/Docks/Forms', ['VSTForm']);
end;

procedure TForm1.FormCreate(Sender: TObject);
begin
  PluginsManager.Loaders.Add
    (TQDLLLoader.Create(ExtractFilePath(Application.ExeName), '.DLL'));
  PluginsManager.Start;
end;

end.
